### PR TITLE
Replacing network ConnectionEvent bitfield with enum.

### LINKED ISSUE #912: clean up Network::ConnectionEvent
Because of libevent, it used to be possible for events to be combined, so Network::ConnectionEvent is a bitfield. That can no longer happen. We should fix it so that it's a real enum which will make a bunch of code simpler and easier to read.

### PR DESCRIPTION
**Pull request by alyssawilk**
Fixes #912

### CODE REVIEW

File: source/common/network/connection_impl.h [hunk 1]
```diff
@@ -86,9 +86,9 @@ class ConnectionImpl : public virtual Connection,
     uint64_t bytes_processed_;
   };
 
-  virtual void closeSocket(uint32_t close_type);
+  virtual void closeSocket(ConnectionEvent close_type);
   void doConnect();
-  void raiseEvents(uint32_t events);
+  void raiseEvents(ConnectionEvent event);
```
Comment (dnoe) [inline source/common/network/connection_impl.h]: Could be renamed raiseEvent to match onEvent (which happened to be wrong before and is correct now :)

File: test/integration/utility.h [hunk 2]
```diff
@@ -110,10 +110,10 @@ class ConnectionStatusCallbacks : public Network::ConnectionCallbacks {
...
 
   // Network::ConnectionCallbacks
-  void onEvent(uint32_t events) override {
-    closed_ |= (events & Network::ConnectionEvent::RemoteClose ||
-                events & Network::ConnectionEvent::LocalClose);
-    connected_ |= events & Network::ConnectionEvent::Connected;
+  void onEvent(Network::ConnectionEvent event) override {
+    closed_ |= (event == Network::ConnectionEvent::RemoteClose ||
+                event == Network::ConnectionEvent::LocalClose);
+    connected_ |= (event == Network::ConnectionEvent::Connected);
```
Comment (dnoe) [inline test/integration/utility.h]: Hmm, isn't |= a bitwise operator? This doesn't look right to me. I'm not sure it was right before, either?

File: include/envoy/network/connection.h [hunk 3]
```diff
@@ -19,13 +19,12 @@ class Dispatcher;
...
 /**
- * Events that occur on a connection. Maybe be combined.
+ * Events that occur on a connection.
  */
-class ConnectionEvent {
-public:
-  static const uint32_t RemoteClose = 0x1;
-  static const uint32_t LocalClose = 0x2;
-  static const uint32_t Connected = 0x4;
+enum ConnectionEvent {
```
Comment (htuch) [inline include/envoy/network/connection.h]: Prefer `enum class`, more type safe.

File: test/integration/utility.h (hunk 2 above)
Comment (htuch) [inline test/integration/utility.h]: I think it's fine if you're using the least significant bit as the boolean value (and are happy with some implicit boolean/integer casting going on under the hood).

File: source/common/network/connection_impl.h (hunk 1 above)
Comment (alyssawilk) [inline source/common/network/connection_impl.h]: Good call! too bad your reward is another 100 lines of string replaces...
Mostly test code though :-)

**Comment by mattklein123** [comment]
@alyssawilk random GH tip: If you put "Fixes #912" in the description, that PR will be auto closed when this merges.

Anyway, this LGTM other than the enum class comment.

File: test/integration/utility.h (hunk 2 above)
Comment (dnoe) [inline test/integration/utility.h]: It looks like the standard concretely guarantees that false is 0 and true is 1, so I think it's safe from UB. I think ye olde time C (before C99) only guaranteed that true expressions were non-zero. I guess it still works in that case but looks odd.

I'm OK with leaving it as is.

File: include/envoy/network/connection.h (hunk 3 above)
Comment (alyssawilk) [inline include/envoy/network/connection.h]: Ooh, good call.  Updated this, and a bunch of tests that subsequently failed to compile.

File: source/common/network/connection_impl.cc [hunk 4]
```diff
@@ -136,7 +136,7 @@ void ConnectionImpl::closeSocket(ConnectionEvent close_type) {
     return;
   }
 
-  ENVOY_CONN_LOG(debug, "closing socket: {}", *this, close_type);
+  ENVOY_CONN_LOG(debug, "closing socket: {}", *this, static_cast<uint32_t>(close_type));
```
Comment (htuch) [inline source/common/network/connection_impl.cc]: Might be worth a `<<` override, up to you. I wish C++ had a better stringification for `eum class` by default.

**Pull request merged**